aloha-core/src/main/scala/com/eharmony/aloha/dataset/vw/multilabel/VwMultilabelRowCreator.scala

```diff
@@ -0,0 +1,361 @@
+  /**
+    * Find the first two valid characters that can be used as VW namespaces that, when converted
+    * to integers are not present in usedNss.
+    * @param usedNss the set of first characters in namespaces.
+    * @return the namespace to use for the actual classes and dummy classes, respectively.
+    */
+  private[multilabel] def bruteForceNsSearch(usedNss: sci.BitSet): Option[LabelNamespaces] = {
+    val found =
+      Iterator
+        .range(Char.MinValue, Char.MaxValue)
+        .filter(c => !(usedNss contains c) && validCharForNamespace(c.toChar))
+        .take(2)
+        .toList
+
+    found match {
+      case actual :: dummy :: Nil =>
+        Option(LabelNamespaces(actual.toChar, dummy.toChar))
+      case _ => None
+    }
+  }
+
+  /**
+    * Produce a multi-line input to be consumed by the underlying ''CSOAA LDF'' VW model.
+    * @param features (non-label dependent) features shared across all labels.
+    * @param indices the indices `labels` into the sequence of all labels encountered
+    *                during training.
+    * @param positiveLabelIndices a predicate telling whether the example should be positively
+    *                             associated with a label.
+    * @param defaultNs the indices into `features` that should be placed in VW's default
+    *                  namespace.
+    * @param namespaces the indices into `features` that should be associated with each
+    *                   namespace.
+    * @param classNs a namespace for features associated with class labels
+    * @param dummyClassNs a namespace for features associated with dummy class labels
+    * @return an array to be passed directly to an underlying `VWActionScoresLearner`.
+    */
+  private[aloha] def trainingInput(
+      features: IndexedSeq[Sparse],
+      indices: sci.IndexedSeq[Int],
+      positiveLabelIndices: Int => Boolean,
+      defaultNs: List[Int],
+      namespaces: List[(String, List[Int])],
+      classNs: Char,
+      negativeDummyStr: String,
+      positiveDummyStr: String
+  ): Array[String] = {
+
+    val n = indices.size
+
+    // The length of the output array is n + 3.
+    //
+    // The first row is the shared features. These are features that are not label dependent.
+    // Then comes two dummy classes.  These are to make the probabilities work out.
+    // Then come the features for each of the n labels.
+    val x = new Array[String](n + 3)
+
+    val shared = VwRowCreator.unlabeledVwInput(features, defaultNs, namespaces, false)
+    x(0) = SharedFeatureIndicator + shared
+
+    // These string interpolations are computed over and over but will always be the same
+    // for a given dummyClassNs.
+    x(1) = negativeDummyStr
+    x(2) = positiveDummyStr
+
+    // This is mutable because we want speed.
+    var i = 0
+    while (i < n) {
+      val labelInd = indices(i)
+
+      // TODO or positives.contains(labelInd)?
+      val dv = if (positiveLabelIndices(i)) PositiveCost else NegativeCost
```

Comment:
So the only problem with the way this is written now is that the function as provided doesn't guarantee an O(1) runtime which you must have for performance.  I would consider changing it so the reader of this function knows that it's an O(1) algorithm.

Reply:
I don't understand.  Are you saying that `positiveLabelIndices` is not guaranteed to be ***O***(**1**)?  That's definitely true, but neither is `Set.contains`.  See [Collections Performance Characteristics](http://docs.scala-lang.org/overviews/collections/performance-characteristics.html).  I don't think this is an issue as it's a package private function that shouldn't be called outside Aloha.